src/app/controllers/myinfo.server.controller.js

```diff
@@ -142,109 +135,93 @@ function _preHashCheckConversion(field) {
   )
   if (authType === 'SP' && actualMyInfoFields.length > 0) {
     const uinFin = res.locals.uinFin
     const formObjId = req.form._id
-
-    let hashedUinFin = crypto
-      .createHmac('sha256', sessionSecret)
-      .update(uinFin)
-      .digest('hex')
-    MyInfoHash.findOne(
-      { uinFin: hashedUinFin, form: formObjId },
-      (err, hashedObj) => {
-        if (err) {
-          logger.error({
-            message: 'Error retrieving MyInfo hash from database',
-            meta: {
-              action: 'verifyMyInfoVals',
-              ...createReqMeta(req),
-            },
-            error: err,
-          })
-          return res.status(StatusCodes.SERVICE_UNAVAILABLE).json({
-            message: 'MyInfo verification unavailable, please try again later.',
-            spcpSubmissionFailure: true,
-          })
+    let hashedFields
+    try {
+      hashedFields = await MyInfoHash.findHashes(uinFin, formObjId)
+    } catch (error) {
+      logger.error({
+        message: 'Error retrieving MyInfo hash from database',
+        meta: {
+          action: 'verifyMyInfoVals',
+          ...createReqMeta(req),
+        },
+        error,
+      })
+      return res.status(StatusCodes.SERVICE_UNAVAILABLE).json({
+        message: 'MyInfo verification unavailable, please try again later.',
+        spcpSubmissionFailure: true,
+      })
+    }
+    if (!hashedFields) {
+      logger.error({
+        message: `Unable to find MyInfo hashes for ${formObjId}`,
+        meta: {
+          action: 'verifyMyInfoVals',
+          ...createReqMeta(req),
+          formId: formObjId,
+        },
+      })
+      return res.status(StatusCodes.GONE).json({
+        message: 'MyInfo verification expired, please refresh and try again.',
+        spcpSubmissionFailure: true,
+      })
+    }
+    // Fields from client submission
+    let clientFormFields = req.body.parsedResponses // responses were transformed in submissions.server.controller.js
+    let clientMyInfoFields = clientFormFields
+      .filter((field) => field.isVisible && field.myInfo && field.myInfo.attr)
+      .map(_preHashCheckConversion)
+
+    // compare hashed values to submission values
+    const bcryptCompares = clientMyInfoFields.map((clientField) => {
+      const expected = hashedFields[clientField.attr]
+      return expected
+        ? bcrypt.compare(clientField.val, expected)
+        : Promise.resolve(true)
+    })
+    Promise.all(bcryptCompares)
+      .then((compare) => {
+        return {
+          compare, // Array<Boolean> indicating hash pass/fail
+          fail: compare.some((v) => !v), // Whether any hashes failed
         }
+      })
+      .then((hashResults) => {
+        if (hashResults.fail) {
+          // Array of MyInfo attributes that failed verification
+          let hashFailedAttrs = _.zip(clientMyInfoFields, hashResults.compare)
+            .filter(([_, compare]) => compare === false)
+            .map(([clientField, _]) => clientField.attr)
```

Comment:
Don't be afraid to extract small functions. A good rule of thumb is to extract a function every time you see a comment that tries to explain a block of code.

Reply:
definitely! I'm just trying to knock out one small one at a time in order to minimise the big one (#560)